### PR TITLE
Feat/google sa expire

### PR DESCRIPTION
**Pull request by paulineribeyre**
Add optional `expires_in` parameter to the POST and PATCH `/google/service_accounts` endpoints (normal and dry_run).
- Check that the provided param is a positive int in `_get_service_account_error_status()` (called by POST, PATCH, POST dry_run and PATCH dry_run).
- Use the provided param in `add_user_service_account_to_db()` (called by POST and PATCH) and `extend_service_account_access` (called by PATCH)

### New Features
- add optional `expires_in` parameter to the POST and PATCH `/google/service_accounts` endpoints (normal and dry_run)
- add optional `expires_in` parameter to `fence-create force-link-google`

### Bug Fixes
- add `expires_in` check to `data/download` and `data/upload` endpoints

### Improvements
- use an util function to check if the provided `expires_in` is a positive int
- add tests for POST and PATCH `/google/service_accounts`
- add tests for GET and PATCH `/link/google`
- make session.pop() `default` param optional

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
This PR contains code that is not formatted correctly according to [`black`](https://github.com/ambv/black). Run `black` on your code before merging.

<details>
<summary>Expand the full diff to see formatting changes</summary>

```diff
--- tests/google/test_service_accounts.py
+++ blackened
@@ -196,26 +196,30 @@
     """
     encoded_creds_jwt = encoded_jwt_service_accounts_access["jwt"]
     service_account = register_user_service_account["service_account"]
 
     # invalid expires_in: should fail
-    requested_exp = "abc" # expires_in must be int >0
+    requested_exp = "abc"  # expires_in must be int >0
     response = client.patch(
-        "/google/service_accounts/{}?expires_in={}".format(quote(service_account.email), requested_exp),
-        headers={"Authorization": "Bearer " + encoded_creds_jwt},
-        content_type="application/json",
-    )
-    assert response.status_code == 400 # check if failure
+        "/google/service_accounts/{}?expires_in={}".format(
+            quote(service_account.email), requested_exp
+        ),
+        headers={"Authorization": "Bearer " + encoded_creds_jwt},
+        content_type="application/json",
+    )
+    assert response.status_code == 400  # check if failure
 
     # valid expires_in: should succeed
     requested_exp = 60
     response = client.patch(
-        "/google/service_accounts/{}?expires_in={}".format(quote(service_account.email), requested_exp),
-        headers={"Authorization": "Bearer " + encoded_creds_jwt},
-        content_type="application/json",
-    )
-    assert str(response.status_code).startswith("2") # check if success
+        "/google/service_accounts/{}?expires_in={}".format(
+            quote(service_account.email), requested_exp
+        ),
+        headers={"Authorization": "Bearer " + encoded_creds_jwt},
+        content_type="application/json",
+    )
+    assert str(response.status_code).startswith("2")  # check if success
 
     # make sure the access was extended of the requested time
     # (allow up to 5 sec for runtime)
     service_account_accesses = (
         db_session.query(ServiceAccountToGoogleBucketAccessGroup).filter_by(
@@ -366,19 +370,21 @@
     """
     encoded_creds_jwt = encoded_jwt_service_accounts_access["jwt"]
     service_account = register_user_service_account["service_account"]
 
     # invalid expires_in: should fail
-    requested_exp = "abc" # expires_in must be int >0
+    requested_exp = "abc"  # expires_in must be int >0
 
     response = client.patch(
-        "/google/service_accounts/_dry_run/{}?expires_in={}".format(quote(service_account.email), requested_exp),
+        "/google/service_accounts/_dry_run/{}?expires_in={}".format(
+            quote(service_account.email), requested_exp
+        ),
         headers={"Authorization": "Bearer " + encoded_creds_jwt},
         content_type="application/json",
         data={"project_access": ["another-valid-project"]},
     )
-    assert response.status_code == 400 # check if failure
+    assert response.status_code == 400  # check if failure
 
     service_account_accesses = (
         db_session.query(ServiceAccountToGoogleBucketAccessGroup).filter_by(
             service_account_id=service_account.id
         )
@@ -391,16 +397,18 @@
 
     # valid expires_in: should succeed
     requested_exp = 60
 
     response = client.patch(
-        "/google/service_accounts/_dry_run/{}?expires_in={}".format(quote(service_account.email), requested_exp),
+        "/google/service_accounts/_dry_run/{}?expires_in={}".format(
+            quote(service_account.email), requested_exp
+        ),
         headers={"Authorization": "Bearer " + encoded_creds_jwt},
         content_type="application/json",
         data={"project_access": ["another-valid-project"]},
     )
-    assert response.status_code == 200 # check if success
+    assert response.status_code == 200  # check if success
 
     service_account_accesses = (
         db_session.query(ServiceAccountToGoogleBucketAccessGroup).filter_by(
             service_account_id=service_account.id
         )
@@ -537,23 +545,23 @@
         headers={"Authorization": "Bearer " + encoded_creds_jwt},
         data=json.dumps(valid_service_account),
         content_type="application/json",
     )
 
-    assert response.status_code == 200 # check if success
+    assert response.status_code == 200  # check if success
 
     # invalid expires_in: should fail
-    requested_exp = "abc" # expires_in must be int >0
+    requested_exp = "abc"  # expires_in must be int >0
 
     response = client.post(
         "/google/service_accounts/_dry_run?expires_in={}".format(requested_exp),
         headers={"Authorization": "Bearer " + encoded_creds_jwt},
         data=json.dumps(valid_service_account),
         content_type="application/json",
     )
 
-    assert response.status_code == 400 # check if failure
+    assert response.status_code == 400  # check if failure
 
 
 def test_invalid_service_account_has_external_access(
     client,
     app,
@@ -915,34 +923,36 @@
     assert len(db_session.query(UserServiceAccount).all()) == 0
     assert len(db_session.query(ServiceAccountAccessPrivilege).all()) == 0
     assert len(db_session.query(ServiceAccountToGoogleBucketAccessGroup).all()) == 0
 
     # invalid expires_in: should fail
-    requested_exp = "abc" # expires_in must be int >0
+    requested_exp = "abc"  # expires_in must be int >0
 
     response = client.post(
         "/google/service_accounts?expires_in={}".format(requested_exp),
         headers={"Authorization": "Bearer " + encoded_creds_jwt},
         data=json.dumps(valid_service_account),
         content_type="application/json",
     )
-    assert response.status_code == 400 # check if failure
+    assert response.status_code == 400  # check if failure
 
     # valid expires_in: should succeed
     requested_exp = 60
 
     response = client.post(
         "/google/service_accounts?expires_in={}".format(requested_exp),
         headers={"Authorization": "Bearer " + encoded_creds_jwt},
         data=json.dumps(valid_service_account),
         content_type="application/json",
     )
-    assert response.status_code == 200 # check if success
+    assert response.status_code == 200  # check if success
 
     assert len(db_session.query(UserServiceAccount).all()) == 1
     assert len(db_session.query(ServiceAccountAccessPrivilege).all()) == 1
-    sa_to_bucket_entries = db_session.query(ServiceAccountToGoogleBucketAccessGroup).all()
+    sa_to_bucket_entries = db_session.query(
+        ServiceAccountToGoogleBucketAccessGroup
+    ).all()
     assert len(sa_to_bucket_entries) == 1
 
     # make sure the access was granted for the requested time
     # (allow up to 5 sec for runtime)
     diff = sa_to_bucket_entries[0].expires - int(time.time())

--- tests/link/test_link.py
+++ blackened
@@ -46,34 +46,28 @@
     """
     encoded_credentials_jwt = encoded_creds_jwt["jwt"]
     redirect = "http://localhost"
 
     # invalid expires_in: should fail
-    requested_exp = "abc" # expires_in must be int >0
+    requested_exp = "abc"  # expires_in must be int >0
 
     r = client.get(
         "/link/google",
-        query_string={
-            "redirect": redirect,
-            "expires_in": requested_exp
-        },
+        query_string={"redirect": redirect, "expires_in": requested_exp},
         headers={"Authorization": "Bearer " + encoded_credentials_jwt},
     )
-    assert r.status_code == 400 # check if failure
+    assert r.status_code == 400  # check if failure
 
     # valid expires_in: should succeed
     requested_exp = 60
 
     r = client.get(
         "/link/google",
-        query_string={
-            "redirect": redirect,
-            "expires_in": requested_exp
-        },
+        query_string={"redirect": redirect, "expires_in": requested_exp},
         headers={"Authorization": "Bearer " + encoded_credentials_jwt},
     )
-    assert r.status_code == 302 # check if success
+    assert r.status_code == 302  # check if success
 
 
 def test_google_link_redirect_no_google_idp(
     client, app, restore_config, encoded_creds_jwt
 ):
@@ -295,22 +289,22 @@
     )
 
     assert not add_google_email_to_proxy_group_mock.called
 
     # invalid expires_in: should fail
-    requested_exp = "abc" # expires_in must be int >0
+    requested_exp = "abc"  # expires_in must be int >0
     r = client.patch(
         "/link/google?expires_in={}".format(requested_exp),
-        headers={"Authorization": "Bearer " + encoded_credentials_jwt}
+        headers={"Authorization": "Bearer " + encoded_credentials_jwt},
     )
     assert r.status_code == 400
 
     # valid expires_in: should succeed
     requested_exp = 60
     r = client.patch(
         "/link/google?expires_in={}".format(requested_exp),
-        headers={"Authorization": "Bearer " + encoded_credentials_jwt}
+        headers={"Authorization": "Bearer " + encoded_credentials_jwt},
     )
     assert r.status_code == 200
 
     account_in_proxy_group = (
         db_session.query(UserGoogleAccountToProxyGroup)
```
</details></br>

This formatting comment was generated automatically by a script in [uc-cdis/wool](https://github.com/uc-cdis/wool).

File: fence/utils.py [hunk 1]
```diff
@@ -252,3 +252,19 @@ def send_email(from_email, to_emails, subject, text, smtp_domain):
         auth=("api", api_key),
         data={"from": from_email, "to": to_emails, "subject": subject, "text": text},
     )
+
+
+def check_expires_in():
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply fence/utils.py]: anyway, lots of words for a minor suggestion. If you disagree and want to keep it, that's cool too

File: fence/scripting/fence_create.py [hunk 2]
```diff
@@ -1292,8 +1292,20 @@ def force_update_google_link(DB, username, google_email):
...
-        now = int(time.time())
-        expiration = now + config["GOOGLE_ACCOUNT_ACCESS_EXPIRES_IN"]
+        # timestamp at which the SA will lose bucket access
+        # by default: use configured time or 7 days
+        expiration = int(time.time()) + config.get(
+            "GOOGLE_USER_SERVICE_ACCOUNT_ACCESS_EXPIRES_IN", 604800
+        )
+        if expires_in:
+            try:
+                expires_in = int(expires_in)
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply fence/scripting/fence_create.py]: `check_expires_in` gets the param from flask arguments, and here it's just a var

[the remark this answers is not in the document]
Comment (paulineribeyre) [reply fence/scripting/fence_create.py]: I could change `check_expires_in` to `check_expires_in(expires_in: None)` and no use flask if the param is provided, if you think that makes more sense

**Pull request merged**